src/rate_limiter.rs

```diff
@@ -5,28 +5,110 @@ use std::time::{Duration, Instant};
 use crate::error::Error;
 use crate::TokenBucket;
 
+/// An object providing rate limiting functionality.
+///
+/// A rate limiter controls how frequently some event, such as an HTTP request,
+/// is allowed to happen. Rate limiting is commonly used as a defensive measure
+/// to protect services from excessive use (intended or not) and maintain their
+/// availability.
+///
+/// A [`RateLimiter`] instance can be used to set how many times an event is
+/// allowed to happen (`limit`) within a given period of time (`interval`). If
+/// no such policy is set for an event, the event is always allowed.
+///
+/// Once constructed, a `RateLimiter` instance is safe to be used from multiple
+/// threads.
+///
+/// Under the hood the token bucket algorithm is used. See [`TokenBucket`] for
+/// details.
+/// 
+/// # Examples
+///
+/// ```
+/// use std::time::Duration;
+/// use youshallnotpass::{RateLimiter, Error};
+///
+/// let limiter = RateLimiter::new()
+///     .limit("A", 2, Duration::from_secs(60))
+///     .limit("B", 3, Duration::from_secs(60))
+///     .build();
+///
+/// assert_eq!(limiter.consume("A", 1), Ok(()));
+/// assert_eq!(limiter.consume("A", 1), Ok(()));
+///
+/// assert!(matches!(limiter.consume("A", 1), Err(Error::RetryAfter(_))));
+/// assert!(matches!(limiter.consume("B", 5), Err(Error::RetryAfter(_))));
+/// ```
 pub struct RateLimiter<'a, K> {
     buckets: HashMap<K, TokenBucket<'a>>,
-    clock: &'a (dyn Fn() -> Instant + Sync),
 }
 
-impl<'a, K: Eq + Hash> RateLimiter<'a, K> {
-    pub fn new() -> Self {
-        Self::with_timer(&Instant::now)
+impl<'a, K> RateLimiter<'a, K> {
+    /// Constructs a new `RateLimiterBuilder` object.
+    ///
+    /// A returned instance of [`RateLimiterBuilder`] can be used to set
+    /// limiting policies and construct an instance of `RateLimiter` via
+    /// [`limit`] and [`build`] functions. See documentation of corresponding
+    /// functions for details.
+    ///
+    /// [`limit`]: RateLimiterBuilder::limit
+    /// [`build`]: RateLimiterBuilder::build
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use youshallnotpass::RateLimiter;
+    ///
+    /// let builder = RateLimiter::<&str>::new();
+    /// ```
+    #[inline]
+    pub fn new() -> RateLimiterBuilder<'a, K> {
```

Comment:
@malor I'm not perfectly happy with `configure()` .. `build()` pair because it looks ugly (at least to someone like me who is not Rustacean). Do you have any other ideas?

I wonder, can we come up with something better than "build"? In case of `File` the building function is called `open()` which looks great.

Reply:
What about `configure()`..`done()` or `with_limits()`..`done()`?